app/src/components/features/rules/RuleBuilder/Header/ActionButtons/CreateRuleButton/actions/insertScriptValidators.js

```diff
@@ -0,0 +1,314 @@
+      validationError: HTMLLinterResult.validationErrors[0].errorType,
+      errorMessage: HTMLLinterResult.validationErrors[0].message,
+    };
+  }
+
+  try {
+    const doc = new DOMParser().parseFromString(str, "text/html");
+
+    if (!isHTMLString(str, doc)) {
+      return {
+        isValid: true,
+      };
+    }
+
+    const nodes = doc.querySelectorAll(htmlNodeName);
+    if (nodes.length === 0) {
+      return {
+        isValid: false,
+        validationError: invalidHTMLError.NO_TAGS,
+        errorMessage: "No tags found in the document",
+      };
+    }
+
+    if (nodes.length > 1) {
+      console.log("3");
+      return {
+        isValid: false,
+        validationError: invalidHTMLError.MULTIPLE_TAGS,
+        errorMessage: `Multiple ${htmlNodeName} tags found in the document. Only one is allowed.`,
+      };
+    }
+
+    if (doesDocumentHaveOtherNodes(doc, htmlNodeName)) {
+      console.log("4", doc);
+      return {
+        isValid: false,
+        validationError: invalidHTMLError.UNSUPPORTED_TAGS,
+        errorMessage: `Only ${htmlNodeName} tags are allowed `,
+      };
+    }
+  } catch (error) {
+    console.debug("couldn't parse the document", error);
+  }
+
+  return {
+    isValid: true,
+  };
+}
+
+/* Pass the code string through an HTML linter and return errors if any */
+async function htmlValidateRawCodeString(codeString) {
+  const loader = new StaticConfigLoader({
+    extends: ["html-validate:recommended"],
+    elements: ["html5"],
+  });
+  const htmlvalidate = new HtmlValidate(loader);
+
+  const report = await htmlvalidate.validateString(codeString);
+
+  if (report.valid) {
+    return {
+      isValid: true,
+      validationErrors: [],
+    };
+  }
+
+  const validationErrors = [];
+
+  /* CURRENTLY WE ARE ONLY CONCERNED WITH SPECIFIC ERRORS */
+
+  report.results.forEach((result) => {
+    if (result.errorCount > 0) {
+      result.messages.forEach((errMessage) => {
+        if (errMessage.ruleId === "parser-error") {
+          validationErrors.push({
+            message: errMessage.message,
+            errorType: invalidHTMLError.COULD_NOT_PARSE,
+          });
+        }
+        if (errMessage.ruleId === "close-order" || errMessage.ruleId === "script-element") {
+          validationErrors.push({
+            message: errMessage.message,
+            errorType: invalidHTMLError.UNCLOSED_TAGS,
+          });
+        }
+        if (errMessage.ruleId === "close-attr") {
+          validationErrors.push({
+            message: errMessage.message,
+            errorType: invalidHTMLError.UNCLOSED_ATTRIBUTES,
+          });
+        }
+      });
+    }
+  });
+
+  return {
+    isValid: validationErrors.length === 0,
+    validationErrors,
+  };
+}
+
+/**
+ * Check if the given string contains HTML nodes
+ * @param {string} str string containing HTML snippet
+ * @param {Document} doc Document created from the HTML string
+ * @returns {boolean}
+ */
+function isHTMLString(str, doc) {
```

Comment:
What is the difference between `isHTMLString` and `doesContainHTMLString` function that you defined above?

https://github.com/requestly/requestly/pull/1386/files#diff-9ba7447af01ba32e558b219199e9e0247e03e17cc3dfdb9e97ed91ccd1e170d9R67

I see the code is being duplicated. Can we avoid it?

Reply:
changed `doesContainHTMLString` to `isJSOrCSS` cause as you mentioned, that was the intent there